src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+            {
+                switch (kind)
+                {
+                    case BlockTypes.Comment: return FoldingRangeKind.Comment;
+                    case BlockTypes.Imports: return FoldingRangeKind.Imports;
+                    case BlockTypes.PreprocessorRegion: return FoldingRangeKind.Region;
+                    default: return null;
+                }
+            }
+        }
+
+        private async Task<VSLocation[]> GetDefinitionAsync(Solution solution, TextDocumentPositionParams request, bool typeOnly, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<VSLocation>.GetInstance();
+
+            var document = solution.GetDocument(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return locations.ToArrayAndFree();
+            }
+
+            var position = await document.GetPositionAsync(request.Position.ToLinePosition(), cancellationToken).ConfigureAwait(false);
+
+            var definitionService = document.Project.LanguageServices.GetService<IGoToDefinitionService>();
+
+            var definitions = await definitionService.FindDefinitionsAsync(document, position, cancellationToken).ConfigureAwait(false);
+            if (definitions != null && definitions.Count() > 0)
+            {
+                foreach (var definition in definitions)
+                {
+                    if (!ShouldInclude(definition))
+                    {
+                        continue;
+                    }
+
+                    var definitionText = await definition.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    locations.Add(new VSLocation
+                    {
+                        Uri = definition.Document.ToUri(),
+                        Range = definition.SourceSpan.ToRange(definitionText),
+                    });
+                }
+            }
+
+            // No definition found - see if we can get metadata as source but that's only applicable for C#\VB.
+            else if (document.SupportsSemanticModel && _metadataAsSourceService != null)
+            {
+                var symbol = await SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).ConfigureAwait(false);
+                if (symbol.Locations.First().IsInMetadata)
+                {
+                    if (!typeOnly || symbol is ITypeSymbol)
+                    {
+                        var declarationFile = await _metadataAsSourceService.GetGeneratedFileAsync(document.Project, symbol, false, cancellationToken).ConfigureAwait(false);
+
+                        var linePosSpan = declarationFile.IdentifierLocation.GetLineSpan().Span;
+                        locations.Add(new VSLocation
+                        {
+                            Uri = new Uri(declarationFile.FilePath),
+                            Range = linePosSpan.ToRange(),
+                        });
```

Comment:
deos thsi work?  if so, that's pretty cool.  how does the external process know things like what project this generate file goes into?

Reply:
current LSP (and live share) doesn't support linked file. that's whole other discussion going on LSP. and LSP VS extension for live share.